Create transparency with respect to the metric’s data and model applicability

annahedstroem opened the pull request:
### Addressed issues:
- https://github.com/understandable-machine-intelligence-lab/Quantus/issues/280
- https://github.com/understandable-machine-intelligence-lab/Quantus/issues/202
- Renaming from `last_results` to `evaluation_scores` and `all_results` to  `all_evaluation_scores`

### Minimum acceptance criteria
- Specify what is necessary for the PR to be merged with the main branch.
- @mentions of the person that is apt to review these changes e.g., @annahedstroem

aaarrti commented:
> ### Addressed issues:
> * [Create transparency with respect to the metric’s individual data and model applicability #280](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/280)
> * [Normalisation per category #202](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/202)
> 
> ### Minimum acceptance criteria
> * Specify what is necessary for the PR to be merged with the main branch.
> * @mentions of the person that is apt to review these changes e.g., @annahedstroem

Just to make sure we are on the same page, the goals of this task are:
- make it easier for users to understand if their data and metric of choice are compatible 
- systematize available metrics, pre-processing features, supported model, etc.

Am I right?

aaarrti commented on quantus/metrics/base.py:
```diff
@@ -101,6 +116,37 @@ def __init__(
+                "Subclasses must define name, data_applicability, model_applicability and score_direction before instantiation"
+            )
+
+    @property
```
Why not just
```python
@property
@abstractmethod
def name(self):
     raise NotImplementedError
```
This way we won't need additional "private" properties, and will enforce inheritors to specify a name.

annahedstroem replied:
Good idea, will implement.

aaarrti commented on quantus/metrics/base.py:
```diff
@@ -18,13 +18,28 @@
     """
     Implementation of the base Metric class.
+
+    Attributes:
```
Typically, attributes, which start with underscore, are supposed to be private, 
meaning, they should not part of the public API, and typically are left out of doc strings. Are you sure they should be mentioned in docstring?

annahedstroem replied:
Hmm good point. But should we instead make them public directly? Why private actually...... this is useful information for the users.

annahedstroem replied:
We can just skip the whole `@property def get_name`  ... by stating them as public attributes straight away.

aaarrti replied:
I'd say completely remove them, since they don't really make sense for base class. Moreover, because of the same name, the descriptors are already defined on the class level, which renders 
```python
@property
@abstractmethod
```
useless. I like the idea of abstract properties much more, since they will enforce each metric to implement them.

annahedstroem replied:
Should be fixed now.

aaarrti commented on quantus/helpers/enums.py:
```diff
@@ -0,0 +1,61 @@
+        - TEXT: Represents text data.
+    """
+
+    IMAGE = "image"  # 3D data
```
`IMAGE`, `TIME-SERIES` and `TABULAR` describe the shapes of the tensors.
While `TEXT` the underlying data type.

Those two look like separate, independent categories to me.
Ideally, we should also specify which data types we expect (or support) `float32`, `float64` or something else?

And finally, for images it would be really useful to make it visible if we expect 
pixel values in the range [0, 255] or [0, 1]. This one always leaves me puzzled 🙃

annahedstroem replied:
> IMAGE, TIME-SERIES and TABULAR describe the shapes of the tensors. While TEXT the underlying data type.

I agree, the text one is vague. I kept it there more as a placeholder until we further define it with an actual usecase.

> Those two look like separate, independent categories to me.

Could be so, should we develop it further when we have the first text model wrapped into Quantus?

> Ideally, we should also specify which data types we expect (or support) float32, float64 or something else?

We should definitely get more precise. Should we define this in a separate PR? I would say that this PR does a first good basic job and then we can go more in detail in a follow-up PR.

> And finally, for images it would be really useful to make it visible if we expect pixel values in the range [0, 255] or [0, 1]. This one always leaves me puzzled 🙃

Hmmm. I think this is out of the scope of Quantus since it is related to the model training process..... e..g, whether the model is trained on [0,1] and not [0, 255] then, ultimately this is also what should be passed as input on test (explanation/ evaluation) time.

aaarrti replied:
Sounds good, I will do a proposal PR with shapes and `dtypes` differentiation on top of this one in the next time then

aaarrti commented on quantus/helpers/enums.py:
```diff
@@ -0,0 +1,61 @@
+from enum import Enum
```
I guess, this is more a question of test. What do you think of using `typing.Literal`? E.g., 

```
from typing import Literal 

DataType = Literal["text", "image", "time-series"]
```

No real benefit, though 🤷‍♂️

annahedstroem replied:
hmm! Literal was introduced in Python3.8, so for our package, enum might be better?

aaarrti replied:
then the question is closed 🤷

aaarrti commented on quantus/helpers/enums.py:
```diff
@@ -0,0 +1,61 @@
+    TEXT = "text"
+
+
+class ModelType(Enum):
```
I'd say `torch` and `tensorflow` describe the underlying framework used for tensor computation, and `ModelType` sound like it is about architecture or task it is solving, so this one is a bit confusing. It would also be great to have some sort of list of supported DNN libraries, since they all have slightly different API, e.g., `torchvision`, `keras`, `transformers`, `keras_cv` ...

annahedstroem replied:
yeah I was thinking that this one could be better developed in the future as we expand the modeltypes (ModelInterface.NewModel) and the list of ModelTypes will basically be those

aaarrti commented on quantus/metrics/base.py:
```diff
@@ -101,6 +116,37 @@ def __init__(
+
+    @property
+    def model_applicability(self):
+        return self._model_applicability
```
I think all metrics support `torch` and `tensorflow`. 
Maybe we could return here both as default, and just in case 1 metric doesn't
we can override the property there.

annahedstroem replied:
for consistency, let's type everything

annahedstroem replied:
especially in the future, we might have less metrics coverage for certain metrics

aaarrti commented on quantus/metrics/robustness/relative_representation_stability.py:
```diff
@@ -22,21 +22,37 @@
     """
-    Relative Output Stability leverages the stability of an explanation with respect
-    to the change in the output logits
+    Relative Representation Stability leverages the stability of an explanation with respect
```
Oh, that's a typo I did 🤦 
Thanks for noticing!

annahedstroem replied:
np!

aaarrti commented on quantus/metrics/base.py:
```diff
@@ -228,6 +274,8 @@ def __call__(
         # Call custom post-processing.
         self.custom_postprocess(**data)
 
+        self.skill_scores = self.compute_skill_score(a_batch=data["a_batch"])
```
Where is `compute_skill_score` method defined?

annahedstroem commented:
> > ### Addressed issues:
> > 
> > * [Create transparency with respect to the metric’s individual data and model applicability #280](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/280)
> > * [Normalisation per category #202](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/202)
> > 
> > ### Minimum acceptance criteria
> > 
> > * Specify what is necessary for the PR to be merged with the main branch.
> > * @mentions of the person that is apt to review these changes e.g., @annahedstroem
> 
> Just to make sure we are on the same page, the goals of this task are:
> 
> * make it easier for users to understand if their data and metric of choice are compatible
> * systematize available metrics, pre-processing features, supported model, etc.
> 
> Am I right?

Yes, this is a starting point for this effort. Then we can complement this PR with a more in-depth ModelType and DataType checker, as the need arise.

aaarrti commented on quantus/metrics/base.py:
```diff
@@ -98,8 +129,28 @@ def __init__(
+    def evaluation_category(self):
+        raise NotImplementedError
+
+    @property
```
This one can safely be implemented in the base class, 
```python
@property
def model_applicability(self) -> Set[ModelType]:
     return {ModelType.Torch, ModelType.TensorFlow}
```
Since all metrics support both, afaik.

annahedstroem replied:
Sure, can be done.

aaarrti commented on quantus/metrics/base.py:
```diff
@@ -98,8 +129,28 @@ def __init__(
+
+    @property
+    @abstractmethod
+    def data_applicability(self):
```
Please add type hints for properties.
